src/msg/async/AsyncMessenger.cc

```diff
@@ -273,6 +276,37 @@ class C_handle_reap : public EventCallback {
   }
 };
 
+/*******************
+ * Admin Socket Hook
+ */
+
+AsyncMessengerSocketHook::AsyncMessengerSocketHook(
+    AdminSocket& asok, AsyncMessenger& m, const std::string& name)
+    : m_asok(asok),
+      m_msgr(m),
+      m_name(name),
+      m_command(fmt::format("messenger dump {}", name)) {}
+
+AsyncMessengerSocketHook::~AsyncMessengerSocketHook() {
+  m_asok.unregister_commands(this);
+}
+
+int AsyncMessengerSocketHook::call(
+    std::string_view command, const cmdmap_t& cmdmap, const bufferlist&,
+    Formatter* f, std::ostream& errss, ceph::buffer::list& out) {
+  if (command == m_command) {
```

Comment:
We definitely don't want to add an actual command for each messenger.

I think you can just add

`std::unique_ptr<AdminSocketHook> CephContext::msgr_hook;`

if not already set (no messenger instance yet associated with the CephContext). Then dynamic cast it to this type to add messenger instances (named) to a map it maintains.

The single command should be able to select a messenger to dump or list the available messengers or dump all of them.

Reply:
Sounds good. The last push adds that unique ptr to the ceph context. However without any mutex. Is it right to assume that messengers will never be initialized concurrently?  (I didn't find any cases where they are.)